apisix/core/config_xds.lua

```diff
@@ -101,18 +128,254 @@ local function load_libxds(lib_name)
               table.concat(tried_paths, '\r\n', 1, #tried_paths))
     end
 
-    local route_zone = C.ngx_http_lua_ffi_shdict_udata_to_zone(route_config[1])
-    local route_shd_cdata = ffi.cast("void*", route_zone)
-    xdsagent.initial(route_shd_cdata)
+    local config_zone = C.ngx_http_lua_ffi_shdict_udata_to_zone(config[1])
+    local config_shd_cdata = ffi.cast("void*", config_zone)
+
+    local conf_ver_zone = C.ngx_http_lua_ffi_shdict_udata_to_zone(conf_ver[1])
+    local conf_ver_shd_cdata = ffi.cast("void*", conf_ver_zone)
+
+    xdsagent.initial(config_shd_cdata, conf_ver_shd_cdata)
+end
+
+
+
+local function sync_data(self)
+    if not latest_version then
+        return false, "wait for more time"
+    end
+
+    if self.conf_version == latest_version then
+        return true
+    end
+
+    if self.values then
+        for _, val in ipairs(self.values) do
+            if val and val.clean_handlers then
+                for _, clean_handler in ipairs(val.clean_handlers) do
+                    clean_handler(val)
+                end
+                val.clean_handlers = nil
+            end
+        end
+        self.values = nil
+        self.values_hash = nil
+    end
+
+    local keys = config:get_keys(0)
+
+    if not keys or #keys <= 0 then
+        -- xds did not write any data to shdict
+        return true, "no keys"
+    end
+
+    -- v1 version we only support route/upstream
+    local capacity = math_ceil(#keys / 2)
```

Comment:
Why divide 2?

Reply:
ngx.shared["xds-config"] stores all routes, upstreams, here I estimate the capacity of the route by using the total capacity / 2.

Reply:
Route and upstream are not 1:1.
Maybe we can use `#keys` directly

Reply:
fix